lib/compress/zstd_compress.c

```diff
@@ -2274,6 +2285,39 @@ static size_t ZSTD_compressBlock_internal(ZSTD_CCtx* zc,
     return cSize;
 }
 
+static size_t ZSTD_compressBlock_targetCBlockSize(ZSTD_CCtx* zc,
+                               void* dst, size_t dstCapacity,
+                               const void* src, size_t srcSize,
+                               U32 lastBlock) {
+    size_t cSize = 0;
+    DEBUGLOG(5, "ZSTD_compressBlock_targetCBlockSize (dstCapacity=%u, dictLimit=%u, nextToUpdate=%u, srcSize=%zu)",
+                (unsigned)dstCapacity, (unsigned)zc->blockState.matchState.window.dictLimit, (unsigned)zc->blockState.matchState.nextToUpdate, srcSize);
+
+    {   const size_t bss = ZSTD_buildSeqStore(zc, src, srcSize);
+        FORWARD_IF_ERROR(bss);
+        if (bss == ZSTDbss_compress) {
+            cSize = ZSTD_compressSuperBlock(zc, dst, dstCapacity, lastBlock);
+    }   }
+
+    if (!ZSTD_isError(cSize) && cSize != 0) {
+        /* confirm repcodes and entropy tables when emitting a compressed block */
+        ZSTD_compressedBlockState_t* const tmp = zc->blockState.prevCBlock;
+        zc->blockState.prevCBlock = zc->blockState.nextCBlock;
+        zc->blockState.nextCBlock = tmp;
+    }
+    /* We check that dictionaries have offset codes available for the first
+     * block. After the first block, the offcode table might not have large
+     * enough codes to represent the offsets in the data.
+     */
+    if (zc->blockState.prevCBlock->entropy.fse.offcode_repeatMode == FSE_repeat_valid)
+        zc->blockState.prevCBlock->entropy.fse.offcode_repeatMode = FSE_repeat_check;
+
+    if (cSize == 0) {
+      cSize = ZSTD_noCompressBlock(dst, dstCapacity, src, srcSize, lastBlock);
```

Comment:
Do we want to emit many small uncompressed blocks instead of one large block?

Reply:
Yes, it seems to match the scenario better :
we want the decoder to regularly decode some output, every packet or so.

When the block is not compressed, it's in theory possible to "stream intra-block",
thus emitting some output without waiting for the full block.
In practice, this ability is not present yet in the reference decoder, 
which reads the block header, determine the block size, and then wait for the full block.

Emitting one big uncompressed block kind of rely on such capability to be developed one day. 
But it doesn't address older versions.